File: src/corelibs/U2Designer/src/dashboard/ParametersWidget.cpp
/**
 * UGENE - Integrated Bioinformatics Tools.
 * Copyright (C) 2008-2013 UniPro <ugene@unipro.ru>
 * http://ugene.unipro.ru
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include "ParametersWidget.h"

#include <U2Core/U2SafePoints.h>

#include <U2Lang/Dataset.h>
#include <U2Lang/URLAttribute.h>
#include <U2Lang/URLContainer.h>
#include <U2Lang/WorkflowUtils.h>


namespace U2 {

ParametersWidget::ParametersWidget(const QWebElement &container, Dashboard *parent)
    : DashboardWidget(container, parent)
{
    const WorkflowMonitor *workflowMonitor = dashboard->monitor();
    SAFE_POINT(NULL != workflowMonitor, "NULL workflow monitor!", );

    const QList<WorkerParamsInfo> &workersParamsInfo = workflowMonitor->getWorkersParameters();
    createWidget(workersParamsInfo);
}

void ParametersWidget::createWidget(const QList<WorkerParamsInfo> &workersParamsInfo) {
    int i = 0;
    foreach (const WorkerParamsInfo &info, workersParamsInfo) {
        // Create the tab
        QString tabId = "params_tab_id_" + QString::number(i);
        QString createTabFunc;
        if (0 == i) {
            createTabFunc = "pwInitAndActiveTab";
        }
        else {
            createTabFunc = "pwAddTab";
        }
        createTabFunc += "(this, '" + info.workerName + "', '" + tabId + "')";

        container.evaluateJavaScript(createTabFunc);

        // Add the parameters
        foreach (Attribute* param, info.parameters) {
            SAFE_POINT(NULL != param, "NULL attribute!", );

            QVariant paramValueVariant = param->getAttributePureValue();
            if (paramValueVariant.canConvert< QList<Dataset> >()) {
                QList<Dataset> sets = paramValueVariant.value< QList<Dataset > >();

                foreach (const Dataset &set, sets) {
                    QString paramName = param->getDisplayName();
                    if (sets.size() > 1) {
                        paramName += ": <i>" + set.getName() + "</i>";
                    }
                    QStringList urls;
                    foreach (URLContainer *c, set.getUrls()) {
                        urls << c->getUrl();
                    }
                    QString paramValue = urls.join(";");

                    QString createParamFunc = "pwAddDatasetParameter";
                    createParamFunc += "('" + tabId + "', ";
                    createParamFunc += "'" + paramName + "', ";
                    createParamFunc += "'" + paramValue + "')";

                    container.evaluateJavaScript(createParamFunc);
                }
            }
            else {
                QString paramName = param->getDisplayName();
                QString paramValue = WorkflowUtils::getStringForParameterDisplayRole(paramValueVariant);

                QString createParamFunc = "pwAddCommonParameter";
                createParamFunc += "('" + tabId + "', ";
                createParamFunc += "'" + paramName + "', ";
                createParamFunc += "'" + paramValue + "')";

                container.evaluateJavaScript(createParamFunc);
            }
        }

        // Increase the iterator
        i++;
    }
}


} // namespace U2

File: src/corelibs/U2Designer/javascript/ParametersWidget.js
/**
 * UGENE - Integrated Bioinformatics Tools.
 * Copyright (C) 2008-2013 UniPro <ugene@unipro.ru>
 * http://ugene.unipro.ru
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

 /** Creates the ParametersWidget layout and the first active tab (without parameters). */
function pwInitAndActiveTab(container, activeTabName, activeTabId) {
    var mainHtml =
        '<div class="tabbable tabs-left">' +
            '<ul class="nav nav-tabs params-nav-tabs">' +
                '<li class="active"><a href="#' + activeTabId + '" data-toggle="tab">' + activeTabName + '</a></li>' +
            '</ul>' +
            '<div class="tab-content params-tab-content">' +
                '<div class="tab-pane active" id="' + activeTabId + '">' +
                    '<table class="table table-bordered table-fixed">' +
                        '<col width="45%">' +
                        '<col width="55%">' +
                        '<thead>' +
                            '<tr>' +
                                '<th><span class="text">Parameter</span></th>' +
                                '<th><span class="text">Value</span></th>' +
                            '</tr>' +
                        '</thead>' +
                        '<tbody scroll="yes">' +
                        '</tbody>' +
                    '</table>' +
                '</div>' +
            '</div>' +
        '</div>';
    
    container.innerHTML = mainHtml;
}

/** Add a parameter with the specified name and value to the specified parameters tab. */
function pwAddCommonParameter(paramsTabId, paramName, paramValue) {
    var paramsTab = document.getElementById(paramsTabId);
    var paramsTable = paramsTab.getElementsByTagName('table')[0];
    var rowCount = paramsTable.rows.length;
    var row = paramsTable.insertRow(rowCount);
    var cell1 = row.insertCell(0);
    var cell2 = row.insertCell(1);
    cell1.innerHTML = wrapLongText(paramName);
    cell2.innerHTML = wrapLongText(paramValue);
}

/**
 * Add a parameter with the specified name and several values.
 * The values should be joined by ';' and
 * input to the method as a single string ('paramValue').
 */
function pwAddDatasetParameter(paramsTabId, paramName, paramValue) {
    var paramsTab = document.getElementById(paramsTabId);
    var paramsTable = paramsTab.getElementsByTagName('table')[0];

    var urls = paramValue.split(';');

    var rowCount = paramsTable.rows.length;
    var row1 = paramsTable.insertRow(rowCount);

    var cell1 = row1.insertCell(0);
    cell1.innerHTML = wrapLongText(paramName);
    cell1.rowSpan = urls.length;

    var cell2 = row1.insertCell(1);
    cell2.innerHTML = urls[0];

    for (var i = 1; i < urls.length; i++) {
        var row = paramsTable.insertRow(rowCount + i);
        var cell = row.insertCell(0);
        cell.innerHTML = urls[i];
    }
}

/**
 * Searches for the ParametersWidget initial layout in the container and
 * appends a new common (non-active) tab without parameters.
 */
function pwAddTab(container, tabName, tabId) {
    var tabsList = container.getElementsByClassName("params-nav-tabs")[0];
    var tabsContent = container.getElementsByClassName("params-tab-content")[0];
    
    var newTabsListEntry = '<li><a href="#' + tabId + '" data-toggle="tab">' + tabName + '</a></li>';
    var newTabsContentEntry =
        '<div class="tab-pane" id="' + tabId + '">' +
            '<table class="table table-bordered table-fixed">' +
                '<col width="45%">' +
                '<col width="55%">' +
                '<thead>' +
                    '<tr>' +
                        '<th><span class="text">Parameter</span></th>' +
                        '<th><span class="text">Value</span></th>' +
                    '</tr>' +
                '</thead>' +
                '<tbody scroll="yes">' +
                '</tbody>' +
            '</table>' +
        '</div>';
    
    tabsList.insertAdjacentHTML('beforeend', newTabsListEntry);
    tabsContent.insertAdjacentHTML('beforeend', newTabsContentEntry);
}
